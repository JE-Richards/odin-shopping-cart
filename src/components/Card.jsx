import PropTypes from 'prop-types';
import { StyledCard } from './styles/Card.styled';

function Card({ obj: { image, price, title } }) {
  return (
    <StyledCard>
      <img src={image} alt={`${title} image.`} />
      <div>
        <h1>{title}</h1>
        <h2>{`$${price}`}</h2>
      </div>
    </StyledCard>
  );
}

Card.propTypes = {
  obj: PropTypes.shape({
    image: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
  }).isRequired,
};

export default Card;
