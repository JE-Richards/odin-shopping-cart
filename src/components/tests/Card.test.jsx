import Card from '../Card';
import { ThemeProvider } from 'styled-components';
import { vi, describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';

const mockTheme = {
  colors: {
    surface: '#ffffff',
    textPrimary: '#000000',
    primary: '#00adb5',
  },
  shadow: '0px 4px 6px rgba(0, 0, 0, 0.1)',
};

describe('Testing the Card component', () => {
  it('correctly renders', () => {
    const testData = {
      image: 'placeholder',
      price: 1000,
      title: 'T-shirt',
    };

    render(
      <ThemeProvider theme={mockTheme}>
        <Card obj={testData} />
      </ThemeProvider>
    );

    const title = screen.getByText(testData.title);
    const price = screen.getByText(`$${testData.price}`);
    expect(title).toBeInTheDocument();
    expect(price).toBeInTheDocument();

    const img = screen.getByRole('img');
    expect(img).toBeInTheDocument();
    expect(img).toHaveAttribute('src', testData.image);
    expect(img).toHaveAttribute('alt', `${testData.title} image.`);
  });

  it('logs a PropTypes warning for invalid props', () => {
    const consoleErrorSpy = vi
      .spyOn(console, 'error')
      .mockImplementation(() => {});

    render(
      <ThemeProvider theme={mockTheme}>
        <Card obj={{ image: 123, price: 'abc', title: 456 }} />
      </ThemeProvider>
    );

    expect(consoleErrorSpy).toHaveBeenCalled();

    // Test the warnings are for prop types
    const calls = consoleErrorSpy.mock.calls.flat();
    const warning = calls.join(' ');
    expect(warning).toContain(
      'Invalid prop `obj.image` of type `number` supplied to `Card`, expected `string`.'
    );
    // Note: subsequent prop type warnings get surpressed to avoid cluttering the console, so only testing that the first has executes. This is sufficient for this project.

    consoleErrorSpy.mockRestore();
  });

  it('matches the snapshot', () => {
    const testData = {
      image: 'placeholder',
      price: 1000,
      title: 'T-shirt',
    };

    const { container } = render(
      <ThemeProvider theme={mockTheme}>
        <Card obj={testData} />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
