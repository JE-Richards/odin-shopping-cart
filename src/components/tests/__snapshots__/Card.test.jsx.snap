// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the Card component > matches the snapshot 1`] = `
<div>
  <div
    class="sc-blHHSb iTUEeP"
  >
    <img
      alt="T-shirt image."
      src="placeholder"
    />
    <div>
      <h1>
        T-shirt
      </h1>
      <h2>
        $1000
      </h2>
    </div>
  </div>
</div>
`;
